server/src/http/broadcaster.rs

```diff
@@ -0,0 +1,190 @@
+struct ClientGroup {
+    clients: Vec<mpsc::Sender<sse::Event>>,
+    filter_set: Query<FeatureFilters>,
+    token: EdgeToken,
+}
+
+#[derive(Default)]
+struct BroadcasterInner {
+    active_connections: HashMap<QueryWrapper, ClientGroup>,
+}
+
+pub struct Broadcaster {
+    inner: Mutex<BroadcasterInner>,
+    features_cache: Arc<DashMap<String, ClientFeatures>>,
+}
+
+impl Broadcaster {
+    /// Constructs new broadcaster and spawns ping loop.
+    pub fn new(features: Arc<DashMap<String, ClientFeatures>>) -> Arc<Self> {
+        let this = Arc::new(Broadcaster {
+            inner: Mutex::new(BroadcasterInner::default()),
+            features_cache: features,
+        });
+
+        Broadcaster::spawn_ping(Arc::clone(&this));
+
+        this
+    }
+
+    /// Pings clients every 30 seconds to see if they are alive and remove them from the broadcast
+    /// list if not.
+    fn spawn_ping(this: Arc<Self>) {
+        tokio::spawn(async move {
+            let mut interval = interval(Duration::from_secs(30));
+
+            loop {
+                interval.tick().await;
+                this.remove_stale_clients().await;
+            }
+        });
+    }
+
+    /// Removes all non-responsive clients from broadcast list.
+    async fn remove_stale_clients(&self) {
+        let active_connections = self.inner.lock().active_connections.clone();
+
+        for (_query, mut group) in active_connections {
+          let mut ok_clients = Vec::new();
+
+          for client in group.clients {
+              if client
+                  .send(sse::Event::Comment("keep-alive".into()))
+                  .await
+                  .is_ok()
+              {
+                  ok_clients.push(client.clone());
+              }
+          }
+
+          group.clients = ok_clients;
+        }
+    }
+
+    /// Registers client with broadcaster, returning an SSE response body.
+    /// The current impl takes the feature set as input and sends it to the client as a connected event.
+    ///
+    /// The commented-out arguments are what we'll need to store per client so
+    /// that we can properly filter / format the feature response when they get
+    /// updates later.
+    pub async fn connect(
+        &self,
+        token: EdgeToken,
+        filter_set: Query<FeatureFilters>,
+        query: unleash_types::client_features::Query,
+        features: Json<ClientFeatures>,
+    ) -> Sse<InfallibleStream<ReceiverStream<sse::Event>>> {
+        let (tx, rx) = mpsc::channel(10);
+
+        self.inner
+            .lock()
+            .active_connections
+            .entry(QueryWrapper {
+                query: query.clone(),
+            })
+            .and_modify(|group| {
+                group.clients.push(tx.clone());
+            })
+            .or_insert(ClientGroup {
+                clients: vec![tx.clone()],
+                filter_set,
+                token,
+            });
+
+        tx.send(
+            sse::Data::new_json(features)
```

Comment:
I checked and `data.into()` can be also expressed as more explicit and verbose `sse::Event::Data(sse::Data::new("unleash-connected"))`  but probably it is less idiomatic in Rust

Reply:
yea, "into" is pretty idiomatic Rust code, so I'm gonna leave this as is.